cmd/snap/cmd_run.go

```diff
@@ -227,6 +230,142 @@ func isReexeced() bool {
+		logger.Noticef("WARNING: %s != %s\n", fin.Name(), xauthPathCan)
+		return "", nil
+	}
+
+	// Only do the migration from /tmp since the real /tmp is not visible for snaps
+	if !strings.HasPrefix(fin.Name(), "/tmp/") {
+		return "", nil
+	}
+
+	// We are performing a Stat() here to make sure that the user can't
+	// steal another user's Xauthority file. Note that while Stat() uses
+	// fstat() on the file descriptor created during Open(), the file might
+	// have changed ownership between the Open() and the Stat(). That's ok
+	// because we aren't trying to block access that the user already has:
+	// if the user has the privileges to chown another user's Xauthority
+	// file, we won't block that since the user can just steal it without
+	// having to use snap run. This code is just to ensure that a user who
+	// doesn't have those privileges can't steal the file via snap run
+	// (also note that the (potentially untrusted) snap isn't running yet).
+	fi, err := fin.Stat()
+	if err != nil {
+		return "", err
+	}
+	sys := fi.Sys()
+	if sys == nil {
+		return "", fmt.Errorf(i18n.G("Can't validate file owner"))
+	}
+	// cheap comparison as the current uid is only available as a string
+	// but it is better to convert the uid from the stat result to a
+	// string than a string into a number.
+	if fmt.Sprintf("%d", sys.(*syscall.Stat_t).Uid) != u.Uid {
+		return "", fmt.Errorf(i18n.G("Xauthority file isn't owned by the current user %s"), u.Uid)
+	}
+
+	targetPath := filepath.Join(baseTargetDir, ".Xauthority")
+
+	// Only validate Xauthority file again when both files don't match
+	// otherwise we can continue using the existing Xauthority file.
+	// This is ok to do here because we aren't trying to protect against
+	// the user changing the Xauthority file in XDG_RUNTIME_DIR outside
+	// of snapd.
+	var fout *os.File
+	if osutil.FileExists(targetPath) {
+		if fout, err = os.Open(targetPath); err != nil {
+			return "", err
+		}
+		defer fout.Close()
+		if osutil.FileStreamsEqual(fin, fout) {
+			return targetPath, nil
+		}
+	}
+
+	// Ensure we're validating the Xauthority file from the beginning
+	if _, err := fin.Seek(int64(os.SEEK_SET), 0); err != nil {
+		return "", err
+	}
+
+	// To guard against setting XAUTHORITY to non-xauth files, check
+	// that we have a valid Xauthority. Specifically, the file must be
+	// parseable as an Xauthority file and not be empty.
+	if err := x11.ValidateXauthorityFromFile(fin); err != nil {
+		logger.Noticef("WARNING: invalid Xauthority file: %s", err)
+		return "", nil
+	}
+
+	if osutil.FileExists(targetPath) {
+		err := os.Remove(targetPath)
+		if err != nil {
+			logger.Noticef("WARNING: failed to remove existing file %s", targetPath)
+		}
+	}
+
+	// Read data from the beginning of the file
+	if _, err = fin.Seek(int64(os.SEEK_SET), 0); err != nil {
+		return "", err
+	}
+
+	fout, err = os.OpenFile(targetPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+	if err != nil {
+		return "", err
+	}
+	defer fout.Close()
```

Comment:
It's a bit weird that you os.Open()d above using fout, deferred the Close() then os.OpenFile()d here using fout and defer this Close(). Perhaps with os.Open(), above, just explicitly Close() so there is no confusion?